caffe2/share/contrib/binaries/caffe2_benchmark/caffe2_benchmark.cc

```diff
@@ -0,0 +1,143 @@
+CAFFE2_DEFINE_string(output, "",
+                     "Output that should be dumped after the execution "
+                     "finishes. If multiple outputs are needed, use comma "
+                     "separated string. If you want to dump everything, pass "
+                     "'*' as the output value.");
+CAFFE2_DEFINE_string(output_folder, "",
+                     "The folder that the output should be written to. This "
+                     "folder must already exist in the file system.");
+CAFFE2_DEFINE_int(warmup, 0, "The number of iterations to warm up.");
+CAFFE2_DEFINE_int(iter, 10, "The number of iterations to run.");
+CAFFE2_DEFINE_bool(run_individual, false, "Whether to benchmark individual operators.");
+
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+int main(int argc, char** argv) {
+  caffe2::GlobalInit(&argc, &argv);
+  caffe2::ShowLogInfoToStderr();
+  unique_ptr<caffe2::Workspace> workspace(new caffe2::Workspace());
+
+  // Run initialization network.
+  caffe2::NetDef net_def;
+  CAFFE_ENFORCE(ReadProtoFromFile(caffe2::FLAGS_init_net, &net_def));
+  CAFFE_ENFORCE(workspace->RunNetOnce(net_def));
+
+  // Load input.
+  if (caffe2::FLAGS_input.size()) {
+    vector<string> input_names = caffe2::split(',', caffe2::FLAGS_input);
+    if (caffe2::FLAGS_input_file.size()) {
+      vector<string> input_files = caffe2::split(',', caffe2::FLAGS_input_file);
+      CAFFE_ENFORCE_EQ(
+          input_names.size(), input_files.size(),
+          "Input name and file should have the same number.");
+      for (int i = 0; i < input_names.size(); ++i) {
+        caffe2::BlobProto blob_proto;
+        CAFFE_ENFORCE(caffe2::ReadProtoFromFile(input_files[i], &blob_proto));
+        workspace->CreateBlob(input_names[i])->Deserialize(blob_proto);
+      }
+    } else if (caffe2::FLAGS_input_dims.size()) {
+      vector<string> input_dims_list = caffe2::split(';', caffe2::FLAGS_input_dims);
+      CAFFE_ENFORCE_EQ(
+          input_names.size(), input_dims_list.size(),
+          "Input name and dims should have the same number of items.");
+      for (int i = 0; i < input_names.size(); ++i) {
+        vector<string> input_dims_str = caffe2::split(',', input_dims_list[i]);
+        vector<int> input_dims;
+        for (const string& s : input_dims_str) {
+          input_dims.push_back(caffe2::stoi(s));
+        }
+        caffe2::TensorCPU* tensor =
+            workspace->GetBlob(input_names[i])->GetMutable<caffe2::TensorCPU>();
+        tensor->Resize(input_dims);
+        tensor->mutable_data<float>();
+      }
+    } else {
+      CAFFE_THROW("You requested input tensors, but neither input_file nor "
+                  "input_dims is set.");
+    }
+  }
+
+  // Run main network.
+  CAFFE_ENFORCE(ReadProtoFromFile(caffe2::FLAGS_net, &net_def));
+  caffe2::NetBase* net = workspace->CreateNet(net_def);
+  CHECK_NOTNULL(net);
+
+  LOG(INFO) << "Starting benchmark.";
+  caffe2::QPLConfig::initSampleRate(
```

Comment:
use of `QPLConfig` seems to be the only differentiating factor between this and `speed_benchmark.cc`.  I would recommend adding an option to use `QPLConfig` in the original file rather than copy it all over to here.

Reply:
The main idea is to use the observer approach to collect the runtime metrics. This way, we can use the same mechanism to collect metrics on mobile apps. It has an added benefit that there is no need to add a benchmarking method whenever a new net type is added. My intention is to replace speed_benchmark with this new mechanism.